// sinon-chrome/out/chrome-api/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});
exports['default'] = create;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _apiConfigJs = require('../api/config.js');

var _apiConfigJs2 = _interopRequireDefault(_apiConfigJs);

var _chromeIndex = require('../chrome/index');

var _chromeIndex2 = _interopRequireDefault(_chromeIndex);

var _factoryEvents = require('../factory/events');

var _factoryEvents2 = _interopRequireDefault(_factoryEvents);

var _factoryStubs = require('../factory/stubs');

var _factoryStubs2 = _interopRequireDefault(_factoryStubs);

var _factoryProperty = require('../factory/property');

var _factoryProperty2 = _interopRequireDefault(_factoryProperty);

var _factoryCache = require('../factory/cache');

var _factoryCache2 = _interopRequireDefault(_factoryCache);

var _pluginsCookies = require('../plugins/cookies');

var _pluginsCookies2 = _interopRequireDefault(_pluginsCookies);

var _pluginsI18n = require('../plugins/i18n');

var _pluginsI18n2 = _interopRequireDefault(_pluginsI18n);

/**
 * Create chrome api mock
 * @returns {Object}
 */

function create() {
    return _lodash2['default'].assign((0, _chromeIndex2['default'])(_apiConfigJs2['default']), ChromeManager);
}

var ChromeManager = {

    /**
     * Install plugin
     * @param {Object} plugin
     * @param {Function} plugin.install
     */
    registerPlugin: function registerPlugin(plugin) {
        plugin.install(this);
    },

    /**
     * Reset mack data
     */
    reset: function reset() {
        _factoryEvents2['default'].reset();
        _factoryStubs2['default'].reset();
    },

    /**
     * Reset mock data and behaviour
     */
    flush: function flush() {
        _factoryCache2['default'].flush();
        _factoryStubs2['default'].flush();
        _factoryEvents2['default'].flush();
        _factoryProperty2['default'].flush();
    },

    /**
     * plugin list
     */
    plugins: {
        CookiePlugin: _pluginsCookies2['default'],
        I18nPlugin: _pluginsI18n2['default']
    }
};
module.exports = exports['default'];
